lib/ES/Repo.pm

```diff
@@ -128,25 +130,36 @@ sub _reference_args {
     return ();
 }
 
+#===================================
+sub add_sub_dir {
+#===================================
+    my ( $self, $branch, $dir ) = @_;
+    $self->{sub_dirs}->{$branch} = $dir;
+    return ();
+}
+
 #===================================
 sub has_changed {
 #===================================
     my $self = shift;
     my ( $title, $branch, $path, $asciidoctor ) = @_;
 
-    my $old
-        = $self->tracker->sha_for_branch( $self->name,
-        $self->_tracker_branch(@_) )
-        or return 1;
+    return 1 if exists $self->{sub_dirs}->{$branch};
 
     local $ENV{GIT_DIR} = $self->git_dir;
-
-    my $new = sha_for($branch)
-        or die "Remote branch <origin/$branch> doesn't exist "
-        . "in repo "
-        . $self->name;
+    my $old = $self->_last_commit_info(@_) or return 1;
+
+    my $new;
+    if ( $self->keep_hash ) {
+        $new = $self->_last_commit(@_);
+    } else {
+        $new = sha_for($branch) or die(
+                "Remote branch <origin/$branch> doesn't exist in repo "
+                . $self->name);
+    }
     $new .= '|asciidoctor' if $asciidoctor;
 
+    return $old ne $new if $self->keep_hash;
```

Comment:
If we're keeping the hash then we don't need to check to see if any of the files that we're using changed so we can just return here.